### DISCUSSION TITLE
Trying to get pip-compile manager working

### DISCUSSION THREAD

**Discussion by MrNaif2018** [renovatebot/renovate#10407] [Request Help]
Hi! Thank you for your quick implementation once again @rarkins!
I know that pip-compile doesn't yet support custom mapping scheme we've got in our repo, but decided to start setting it up
In our repo we have configure renovate PR pending, so I am changing config and initially testing the detecting of package files. Not all files were detected, so it might be even an issue with existing `pip_requirements` manager?
PR here: https://github.com/bitcartcc/bitcart/pull/227
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "local>bitcartcc/renovate-config"
  ],
  "pip-compile": {
    "fileMatch": [
      "^requirements(\/.*)\\.txt$"
    ],
    "ignorePaths": [
      "requirements/deterministic/**"
    ]
  }
}
```

For now I have configured a `fileMatch` to fetch all .txt files in requirements folder, excluding actual output files. But instead of 10 files, it has parsed only 3. In the logs:
```
Matched 10 file(s) for manager pip-compile: requirements/base.txt, requirements/daemons/bch.txt, requirements/daemons/bsty.txt, requirements/daemons/btc.txt, requirements/daemons/gzro.txt, requirements/daemons/ltc.txt, requirements/dev.txt, requirements/production.txt, requirements/test.txt, requirements/web.txt
DEBUG: Found circleci package files
DEBUG: Found pip-compile package files
DEBUG: Found 4 package file(s)
INFO: Dependency extraction complete
{
  "baseBranch": "master",
  "stats": {
    "managers": {
      "circleci": {
        "fileCount": 1,
        "depCount": 3
      },
      "pip-compile": {
        "fileCount": 3,
        "depCount": 3
      }
    },
    "total": {
      "fileCount": 4,
      "depCount": 6
    }
  }
}
```

The only files detected are [requirements/daemons/bch.txt](https://github.com/bitcartcc/bitcart/blob/master/requirements/daemons/bch.txt), [requirements/daemons/ltc.txt](https://github.com/bitcartcc/bitcart/blob/master/requirements/daemons/ltc.txt), [requirements/test.txt](https://github.com/bitcartcc/bitcart/blob/master/requirements/test.txt)

From investigating what all those files have in common, I see that all them have one dependency pinned with ranges, others unpinned. Is that an issue or something with my config? I have `:preserveSemverRanges` in my base renovate preset

If possible, we can discuss my attempts at making `pip-compile` working there, till it gets to a working state (:

**Comment by rarkins** [comment]
Some Renovate concepts to understand:
- The files you match for a manager are referred to as "package files"
- Other files which are derived from package files are referred to as lock files or more generally "artifacts"
- Renovate typically doesn't update artifact files directly but instead updates them whenever the package files need updating

For `pip-compile`, we are treating the input files as package files and the output files as artifacts. Hence it only does work when the input file needs updating, and if you had an input file with no constraints then it would never need updating.

Also historically Renovate has ignored dependencies with no constraints, however we could reconsider that for `pip-compile`.

Renovate also has a concept of "lock file maintenance" which essentially means "deleting the artifact file and regenerating it, even if the package file hasn't changed". This is perhaps what you were expecting with `pip-compile`. It wasn't initially enabled but I added it just now in #10408 and it will be live in the app sometime later this weekend.

FYI here is the example repo I tested with: https://github.com/renovate-tests/pip-compile-1

**Reply by MrNaif2018** [reply]
Hmm, I see. I think indeed it needs some changes. How pip-compile usually works is, the input files define a list of your dependencies without any versions (just a list of them), unless you have some strict requirements because of package incompatibility or something. Then you run pip-compile and you basically get both a "lockfile" and a "package.json" in one file.
Then to upgrade some dependency manually we would run:
`pip-compile requirements.in -o requirements.txt --upgrade-package django==3.1`
It would edit the output file and therefore upgrade the dependency. 
https://github.com/jazzband/pip-tools/blob/master/README.rst#updating-requirements
That's the point that the output files are both specifying version and work as a lockfile.
So maybe for pip-compile input files should be allowed to work even without constraints
Because in your sample repo, if you replace `django<3.1` to `django` in your requirements.in (sometimes that's the case that you don't have any hard constraints on a package version in the whole file, in our case it's so because we split requirements into many files by category), then it won't work, as exactly it didn't detect the file in my case.

And so in general not sure how to make this work, but when some package needs upgrading, we would check the output "lockfile" and "package.json" and see if pinned version can be upgraded.
So actually thinking, maybe in pip-compile case package file and artifact is the same file? And to get input file argument to pass to CLI it would be back linked?

Maybe I am telling something wrong, but that's how I understand pip-compile works after using it for a while plus checking some other repositories using it

When I didn't add `ignorePaths` renovate indeed detected my files and even things to upgrade, maybe should I remove them to test it again?
![изображение](https://user-images.githubusercontent.com/39452697/121775398-0be1ab00-cb90-11eb-9f02-c4db428b5569.png)

UPD: or maybe `pip-compile` manager is a combination of pip_requirements manager for package files (.in files) and artifacts (.txt files), plus a bit more additional features for output file generation? As sometimes indeed input files need updating when they have hard pins, and output files too, just that `pip-compile` has a bit more features

UPD2: I think that it could work like so, not sure if it is doable:
`pip-compile` manager settings could be left the same, just with that ability to configure input/output mapping in a better way, for now it is fine.
When enabled, fileMatch matches package files (.in) files, but it will be a special case: if it contains some constraints, then it would be included for update process, maybe via `pip_requirements` utilities. If it does not have constraints, it won't be excluded still. Then, for each input file added, an output file would be got via some transformation, and checked. If it exists, it will be added for `pip-compile` manager upgrades. Also if input file was updated via any way, `pip-compile input-file -o output-file` would need to be re-run too. When only output file needs changing, `pip-compile input-file -o output-file --upgrade package` could be used. That's probably the right way

**Reply by rarkins** [reply]
> Hmm, I see. I think indeed it needs some changes. How pip-compile usually works is, the input files define a list of your dependencies without any versions (just a list of them), unless you have some strict requirements because of package incompatibility or something. Then you run pip-compile and you basically get both a "lockfile" and a "package.json" in one file.

Thank you for the info. I disagree philosophically that it's the best approach for mature dependency management for *applications*. In other words I believe:
- For libraries where de-duplication of transitive dependencies is essential or highly desirable, use ranges
- Otherwise, pin dependencies narrowly for highest control and visibility

I think that the output file of `pip-compile` is still closer to a lockfile than lockfile + package file. The input file is still like a package file even if the *convention* is for user to loosely constrain dependencies. You can do that in `package.json` too but it's not considered good practice.

Anyway, the good news is that this fits within our model.

> Then to upgrade some dependency manually we would run:
> `pip-compile requirements.in -o requirements.txt --upgrade-package django==3.1`
> It would edit the output file and therefore upgrade the dependency.

Copy and pasting from the `pip-tools` readme:

``pip-compile`` generates a ``requirements.txt`` file using the latest versions
that fulfil the dependencies of ``setup.py`` or ``requirements.in``.

If ``pip-compile`` finds an existing ``requirements.txt`` file that fulfils the
dependencies then no changes will be made, even if updates are available.

To force ``pip-compile`` to update all packages in an existing
``requirements.txt``, run ``pip-compile --upgrade``.

To update a specific package to the latest or a specific version use the
``--upgrade-package`` or ``-P`` flag:

```bash
    # only update the django package
    $ pip-compile --upgrade-package django

    # update both the django and requests packages
    $ pip-compile --upgrade-package django --upgrade-package requests

    # update the django package to the latest, and requests to v2.0.0
    $ pip-compile --upgrade-package django --upgrade-package requests==2.0.0
```

End copy/paste.

I guess that `pip-tools --upgrade` is the same as deleting the output file before running `pip-tools`?

In Renovate we generally don't do "unbound" updates so therefore commands like `pip-compile --upgrade-package requests==2.0.0` will be most useful.

> So maybe for pip-compile input files should be allowed to work even without constraints

Yes, I agree

> Because in your sample repo, if you replace `django<3.1` to `django` in your requirements.in (sometimes that's the case that you don't have any hard constraints on a package version in the whole file, in our case it's so because we split requirements into many files by category), then it won't work, as exactly it didn't detect the file in my case.

I think the problem is that it detects zero dependencies with constraints, which counts as zero dependencies, which in turn is classified as an ignored package file.

> And so in general not sure how to make this work, but when some package needs upgrading, we would check the output "lockfile" and "package.json" and see if pinned version can be upgraded.

Here's how I think it can work:
- We need to stop ignoring unconstrained package dependencies
- We need to parse the output file and treat it as the lock file and for each dependency in the input file we record the "lockedVersion"
- We support both `rangeStrategy=replace`, which is the current behavior meaning updates are only made if the input file needs it *as well as* `rangeStrategy=update-lock-file`, which means that if the `lockedVersion` is not the latest then we use a command like `pip-compile --upgrade-package foo=1.2.3` to update the locked version

> When I didn't add `ignorePaths` renovate indeed detected my files and even things to upgrade, maybe should I remove them to test it again?
> ![изображение](https://user-images.githubusercontent.com/39452697/121775398-0be1ab00-cb90-11eb-9f02-c4db428b5569.png)

I think this is the `pip_requirements` manager finding these. I added #10412 for the purpose of automatically ignoring them, but first want to think if that's the best way.

> UPD2: I think that it could work like so, not sure if it is doable:
> `pip-compile` manager settings could be left the same, just with that ability to configure input/output mapping in a better way, for now it is fine.
> When enabled, fileMatch matches package files (.in) files, but it will be a special case: if it contains some constraints, then it would be included for update process, maybe via `pip_requirements` utilities. If it does not have constraints, it won't be excluded still. Then, for each input file added, an output file would be got via some transformation, and checked. If it exists, it will be added for `pip-compile` manager upgrades. Also if input file was updated via any way, `pip-compile input-file -o output-file` would need to be re-run too. When only output file needs changing, `pip-compile input-file -o output-file --upgrade package` could be used. That's probably the right way

Yes, I think that's essentially the same as I described.

But I'm also thinking of another way.

Could we possibly use `fileMatch` to match the *output* files instead of the input ones? Because it seems like the output files already contain lines like this:

```
djangorestframework==3.12.4
    # via -r requirements.in
```

I notice in your bitcart example the upper lines are:

```
# Python version: Python 3.7.10
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile --allow-unsafe --generate-hashes --output-file=requirements/deterministic/web.txt requirements/web.txt
#
```

Out of interest, do you know why the Python version is printed there and wasn't in my example repository?

So perhaps we can use that embedded metadata to learn the "input" file from the "output" file, and then we don't need any further configuration or mapping added anywhere?

Then we could perhaps combine this with the existing `pip_requirements` manager like so:
- Renovate provides the list of all matching files to the `pip_requirements` manager, which parses them all, including if no constraints
- If any of the parsed files are "autogenerated by pip-compile" and we can locate the corresponding input file, we consider the input file to be the package file and not the autogenerated file
- If the input file is not in the existing list of matched files, we find it anyway and return it as a result

Then it would mean the result would be:
- A list of input files, with their dependencies, and locked versions
- Any non auto-generated requirements.txt files too, where there will be no locked versions

What do you think?

**Reply by MrNaif2018** [reply]
About the philosophy of input file conventions: well, in Python world it is somehow made so that people often use plain dependency names instead of ranges, and use external tools to control it if needed for example. I mean, from my experience packages break way less than in JS world (:

 > I guess that pip-tools --upgrade is the same as deleting the output file before running pip-tools?
 
 Yes, see https://github.com/jazzband/pip-tools/blob/f031f69fb6e052711d1d787bfcf473ab501b500a/piptools/scripts/compile.py#L337
 
 Basically, in all cases it rewrites the file, just that if `--upgrade` is not passed and file exists it parses initial file first and bases further changes on that
 
 So we came to the same conclusion on how to handle it, great.
 
 > I think this is the pip_requirements manager finding these. I added #10412 for the purpose of automatically ignoring them, but first want to think if that's the best way.
 
 I don't think so. In my PR description it has shown:
 
![изображение](https://user-images.githubusercontent.com/39452697/121780624-e06bba00-cba9-11eb-82ec-4d1637d6def8.png)
Plus the `fileMatch` was configured only for `pip-compile` manager.
```
INFO: Dependency extraction complete
{
  "baseBranch": "master",
  "stats": {
    "managers": {
      "circleci": {
        "fileCount": 1,
        "depCount": 3
      },
      "pip-compile": {
        "fileCount": 13,
        "depCount": 247
      }
    },
    "total": {
      "fileCount": 14,
      "depCount": 250
    }
  }
}
```
It's job ID 375438714 on github app
The implementation in the PR is fine, but I don't think it is the actual issue.

About the second approach, yeah, I mentioned it too. It might be even better, the only needed thing is a correct parser of that format (and some checks, what happens if file is invalid for example?)
https://github.com/bitcartcc/bitcart/blob/master/requirements/deterministic/web.txt
Take a look at that file, let's say line 491
pynacl says it is added via paramiko, then we recurse to paramiko definition and it says via `-r requirements/web.txt`. If `-r` is there it means read the file.
Also note https://github.com/bitcartcc/bitcart/blob/master/requirements/deterministic/web.txt#L163, one dependency may be required by multiple others, then it is added in multiple lines

This approach is easier if the parser works correctly. Not sure which approach in general is better.
> Out of interest, do you know why the Python version is printed there and wasn't in my example repository?

That's added by my custom `scripts/compile-requirements.sh` script. We need python version there for clarity as it matters a lot in such files.
https://github.com/bitcartcc/bitcart/blob/master/scripts/compile-requirements.sh#L23
This is pretty useful, so maybe either renovate could utilize it too, or it could be added to upstream `pip-tools`.

> Any non auto-generated requirements.txt files too, where there will be no locked versions

This is probably to be delegated to `pip_requirements` and not `pip-compile`, right?

Well yes, all files that were matched from that back-linking technology from `autogenerated by pip-compile` files will be special and used with `pip-compile`, and any other file is then not `pip-compile` one but `pip_requirements`.

So as a conclusion, if the process of finding the file connected to dependency is not hard to implement (especially parsing those commands and edge cases with invalid files), then the second approach is easier, but in general the two approaches share a lot in common, the main difference is how we locate input/output file as far as I've understood

https://github.com/jazzband/pip-tools/blob/f031f69fb6e052711d1d787bfcf473ab501b500a/piptools/writer.py#L247 pip-tools parses requirements file by `pip` directly, and then uses it's attributes to get the source. Maybe the same could be utilized, if using pip directly is fine for renovate


UPD: https://github.com/jazzband/pip-tools/commit/c7e6d849813fd595e8026894d5efa217a3254a89

Great, python version is now logged, just not in a release yet

UPD2: https://github.com/jazzband/pip-tools/pull/1419
Package ordering issue might be resolved, or is it related to something else?

**Reply by craigds** [reply]
```
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile --allow-unsafe --generate-hashes --output-file=requirements/deterministic/web.txt requirements/web.txt
```

> So perhaps we can use that embedded metadata to learn the "input" file from the "output" file, and then we don't need any further configuration or mapping added anywhere?

Parsing that commet sounds like a good enhancement, but note that the env var `CUSTOM_COMPILE_COMMAND` can affect this header, and is used to point to people's custom wrapper scripts. Plus the specific output is likely to vary occasionally between pip-compile versions. So it'd be best to use current behaviour (assume `x.in` --> `x.txt`) as a fallback I think. Or provide a way to specifically configure the mapping...

> Could we possibly use fileMatch to match the output files instead of the input ones? Because it seems like the output files already contain lines like this:
```

djangorestframework==3.12.4
   # via -r requirements.in
```

I like this suggestion; but note that it's a bit more complex; the comment can also point to a `.txt` file when multiple .in files exist with dependencies between them:


## production.in 
```
django-rest-framework
```

## production.txt
```
#...
django-rest-framework==0.1.0
    # via -r ./production.in
```

## local.in
```
-r production.txt
pytest
```

## local.txt
```
#...
django-rest-framework==0.1.0
    # via -r ./production.txt
pytest==6.2.4
    # via -r ./local.in
```

(Additionally, `local.in` _could_ actually include other `.in` files directly as well (this is discouraged but someone out there probably does it), leading to a situation where `local.txt` contains comments pointing to multiple `.in` files rather than just one.)

If we ignore that particular oddball situation, I guess that for a given .txt file, you can determine which `.in` file to compile by just looking for these comments and ignoring ones that mention a `.txt` file. Then hopefully you end up with one specific `.in` file which can be compile to produce that `.txt` file.


(Since I'm mentioning about files including each other, am I right in saying that renovate's recently-merged pip-compile support doesn't yet include support for file includes? The original ticket contained a lot of discussion about building up graphs to figure out what order to compile files in, but I don't see that reflected in the code yet.)

thanks for working on this, it's exciting :)

**Reply by rarkins** [reply]
Sorry cannot process everything you wrote yet but can confirm that we don't yet build up any graph/dependency list. I think that if we sorted the "extracted" list of returned dependency files based on this dependency ordering then the problem would kind of solve itself as that's the same order they'd be processed in during the update phase.

**Reply by rarkins** [reply]
I would prefer that we base our command off the comments like `pip-compile --allow-unsafe --generate-hashes --output-file=requirements/deterministic/web.txt requirements/web.txt` instead of the env variable `CUSTOM_COMPILE_COMMAND` so that behavior is more predictable. Would it be a *reasonable* restrictions to say that if you want to use Renovate with pip-compile then you need to avoid using the env variable, or are there use cases for where that would be an unreasonable requirement?

**Reply by craigds** [reply]
Yes it's probably reasonable to not support usage of that env var, if it makes the pip-compile args detection simpler.

**Reply by terencehonles** [reply]
Sorry if this is bumping an old thread or is hijacking it for a different issue. I'm also trying to get the `pip-compile` manager working. I raised an issue previously with pip-tools (https://github.com/jazzband/pip-tools/issues/1365#issuecomment-902269288) and replied more recently with its relation to renovate and the issue I was running into. Part of the underlying issue is that pip-compile doesn't have its own resolver and it can run into issues relatively easily.

I have pulled out the config into a test repository and this is the pull request I am having an issue with https://github.com/terencehonles/test-renovate/pull/1

The problem is that renovate is only updating the `requirements.in` file and not the `requirements.txt` file. I optimistically tried the range strategy `update-lockfile` since there was some discussion above by @rarkins, but it looks like that has not been implemented (the logs said it was unknown and `replace` would be used instead). The reason this runs into issues is because as mentioned in https://github.com/jazzband/pip-tools/issues/1365#issuecomment-902269288 pip-compile will use the `requirements.txt` to select the package version to use even though it's no longer selected in `requirements.in`. As mentioned in the comment you can use `--upgrade` or potentially another option, but I'm not sure if that's appropriate in this context because it may be upgrading too many packages.

**Reply by MrNaif2018** [reply]
@terencehonles `pip-compile` support is unfinished and currently isn't very usable. I plan to implement it properly soon, but I suggest to move to #2334.
Here's how I will probably implement it: https://github.com/renovatebot/renovate/issues/2334#issuecomment-915523842

**Reply by terencehonles** [reply]
I'm already in that thread too :sweat_smile:. I'm not sure where this discussion _should_ be happening, but I'm describing a much simpler case than you're describing and I would expect what I have to have at least worked with the initial implementation. I understand at a high level why it does not, but in the short term I _think_ the `pip-compile` manager could be configured to bump version constraints in both `requirements.in` and `requirements.txt` and I _think_ that would allow it to work. I've not tested this, and it's quite possible the command would need to include `--upgrade-package` as you mention, but I'm not sure the argument's behavior for transitive dependencies.
